mjolnir: init at 1.1.20, nixos/mjolnir: init, nixos/pantalaimon: init

jojosch opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Added [mjolnir](https://github.com/matrix-org/mjolnir), a matrix moderation tool. Can be used together with #123879.

I also added a service definition for `pantalaimon` to be able to join mjolnir to encrypted rooms (which is [recommended by upstream](https://github.com/matrix-org/mjolnir#bot-configuration) for at least the "moderator room").

I copied the descriptions for the various options from upstream.

Example config with pantalaimon:

```nix
{ config, pkgs, ... }: {
  services.mjolnir = {
    enable = true;
    homeserverUrl = "https://matrix.domain.tld";
    #accessTokenFile = "/run/secrets/mjolnir-token";
    pantalaimon = {
      enable = true;
      username = "mjolnir";
      passwordFile = "/run/secrets/mjolnir-password";
    };
    protectedRooms = [
      "https://matrix.to/#/!xxx:domain.tld"
    ];
    managementRoom = "!yyy:domain.tld";
    settings = {
      autojoinOnlyIfManager = true;
      recordIgnoredInvites = true;
      acceptInvitesFromGroup = "+moderators:domain.tld";
    };
  };
}

```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jojosch committed:
mjolnir: init at 0.1.17

r-rmcgibbo commented:
Result of `nixpkgs-review pr 123896` at 402e44ae run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

symphorien commented:
Pantalaimon has two possible use cases at least: for bots, and for users of non e2e ready software like fractal. For the second use case, pantalaimon is meant to be run as a systemd user service inside the session of the user so that it can display notifications. I am afraid that if a user sees that there is a module for pantalaimon, they try to use it, but it will fail to work because this module uses a system service run as a dynamic user. So I'm not very happy with this module.


Anyway I'm a bit puzzled by this recommendation of e2e room for the bot, as the system administrator has about the same priviledged as mjolnir, right?

aanderse commented on nixos/modules/services/misc/pantalaimon.nix:
```diff
@@ -0,0 +1,173 @@
+          };
+
+          listenPort = mkOption {
+            type = types.int;
```
`types.port`

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+      '';
+    };
+
+    accessToken = mkOption {
```
We don't need this option. If someone is really motivated to store secrets in the nix store they are free to set the value of `accessTokenFile` to something like `pkgs.writeTextFile "badidea.txt" "notsosecret"`.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+            description = "The username to login with.";
+          };
+
+          password = mkOption {
```
See above comment.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+      serviceConfig = {
+        ExecStart = ''${pkgs.mjolnir}/bin/mjolnir'';
+        ExecStartPre = mkIf (cfg.pantalaimon.enable) [
+          # pantalaimon needs a few seconds to start ...
```
This comment makes me wonder if the `pantalaimon` service is properly configured or not... this can't cleanly be expressed via `systemd` dependencies?

To be clear I'm just double checking. I'm not familiar with the software and sometimes that is the case.

jojosch replied:
I had the situation, that on a new boot (or when pantalaimon wasn't running) the startup failed but the service did not exit with an error code > 0. With `Restart=always` and `RestartSec=5`:

```
[INFO] [index] Starting bot...
[DEBUG] [MatrixHttpClient (REQ-1)] GET http://localhost:8009/_matrix/client/r0/joined_rooms
[ERROR] [MatrixHttpClient (REQ-1)] Error: connect ECONNREFUSED 127.0.0.1:8009
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 8009
}
(node:27088) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:8009
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:27088) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:27088) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
mjolnir.service: Succeeded.
mjolnir.service: Scheduled restart job, restart counter is at 1.
Stopped mjolnir - a moderation tool for Matrix.
Starting mjolnir - a moderation tool for Matrix...
Started mjolnir - a moderation tool for Matrix.
[INFO] [index] Starting bot...
[DEBUG] [MatrixHttpClient (REQ-1)] GET http://localhost:8009/_matrix/client/r0/joined_rooms
[INFO] [index] Resolving protected rooms...
[INFO] [index] Resolving management room...
...
```

Thats also why I've added the `network-online.target`.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+        "null;"
+    }
+
+    module.exports = {
```
You might consider creating a `settings` option where the user can set any value they want. You could then write that to `json` in here. It would add some nice flexibility to the module and reduce the number of options required.

jojosch replied:
I'm now defining only the basic settings which need to be set to use the basic functionality. All other settings can be set via a `settings` attribute.

With this change I could replace the custom patch to load a config from a path from the environment by placing a merged config of

- the default.yaml of mjolnir
- the basic settings defined by the module

in `config/default.yaml`.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+
+    pantalaimon = mkOption {
+      description = ''
+        Pantalaimon options (enables E2E Encryption support)
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+      type = types.path;
+      default = "/var/lib/mjolnir";
+      description = ''
+        The directory the bot should store various bits of information in
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+        ]
+      '';
+      description = ''
+        A list of rooms to protect (matrix.to URLs)
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+    };
+
+    commands = mkOption {
+      description = "Misc options for command handling and commands";
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+            default = true;
+            description = ''
+              If true, ban commands that use wildcard characters require confirmation with
+              an extra `--force` argument
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+    };
+
+    protections = mkOption {
+      description = "Configuration specific to certain toggleable protections";
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+            description = ''
+              How long after a user joins the server should the bot monitor their messages.  After
+              this time, users can say words from the wordlist without being banned automatically.
+              Set to zero to disable (users will always be banned if they say a bad word)
```
Missing period.

aanderse commented on nixos/modules/services/misc/mjolnir.nix:
```diff
@@ -0,0 +1,438 @@
+    systemd.services.mjolnir = {
+      description = "mjolnir - a moderation tool for Matrix";
+      wants = [ "network-online.target" ] ++ optionals (cfg.pantalaimon.enable) [ "pantalaimon.service" ];
+      after = [ "network-online.target" ] ++ optionals (cfg.pantalaimon.enable) [ "pantalaimon.service" ];
```
`network-online.target` or `network.target`? 🤔

aanderse commented on nixos/modules/services/misc/pantalaimon.nix:
```diff
@@ -0,0 +1,173 @@
+          enableDbusUi = false;
+        };
+      '';
+      description = "Package that should be used for pantalaimon (e.g to turn off dbus)";
```
Missing period.

aanderse commented on nixos/modules/services/misc/pantalaimon.nix:
```diff
@@ -0,0 +1,173 @@
+    systemd.services.pantalaimon = {
+      description = "pantalaimon - E2EE aware proxy daemon for matrix clients";
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
```
`network-online.target` or `network.target`? 🤔

aanderse reviewed (commented):
I left some comments on the module. Hopefully they are more helpful than not.

jojosch commented:
@symphorien I've added the option to define multiple instances of pantalaimon. With the `headless` option enabled pantalaimon is compatible with bots (like mjolnir) and when the option is disabled a systemd-user service is available.

I've tested both variants (headless "bot" mode with mjolnir and the regular mode on my laptop with fractal as a client). Fractal was able to display messages in encrypted rooms. Verification with `panctl` worked out of the box.

When enabling pantalaimon in mjolnir, a new pantalaimon instance "mjolnir" is created and configured for headless-mode.

jojosch commented:
sorry for that ping - messed up that rebase.

symphorien commented:
If I set up pantalaimon in non-headless mode, and two users login, I think two pantalaimon user services will start up and attempt to open the same port. I don't see any nice solution right now.

sumnerevans commented on nixos/modules/services/misc/pantalaimon.nix:
```diff
@@ -0,0 +1,69 @@
+let
+  cfg = config.services.pantalaimon;
+
+  mkConfigFile = name: instanceConfig: pkgs.writeText "pantalaimon.conf" ''
+    [Default]
+    LogLevel = ${instanceConfig.logLevel}
+    Notifications = ${if instanceConfig.notifications then "On" else "Off"}
+
+    [${name}]
+    Homeserver = ${instanceConfig.homeserver}
+    ListenAddress = ${instanceConfig.listenAddress}
+    ListenPort = ${toString instanceConfig.listenPort}
+    SSL = ${boolToString instanceConfig.ssl}
+    IgnoreVerification = ${boolToString instanceConfig.ignoreVerification}
+    UseKeyring = ${if instanceConfig.useKeyring then "Yes" else "No"}
+    ${optionalString (instanceConfig.proxy != null) ''Proxy = ${instanceConfig.proxy}''}
+
+    ${instanceConfig.extraConfig}
+  '';
```
Maybe this can use `pkgs.formats.ini` (or a modification thereof)?

jojosch replied:
Using `pkgs.formats.ini` now.

sumnerevans reviewed (changes requested):
I think it would be good to add a module test. I worked on creating a pantalaimon module last night (forgot to search for existing ones :facepalm:) and added the following test:

https://github.com/sumnerevans/nixpkgs/blob/3a54b693f8078755afbf54305e26a2bd18e9fbc8/nixos/tests/matrix/pantalaimon.nix

Also, I really think it would be good to add a new category under `modules/services` for `matrix`. I have a feeling that we will be seeing a lot more matrix-related modules soon.

jojosch commented:
I've added the pantalaimon test by @sumnerevans and adapted it to the slightly different service structure. Also moved the modules into a "matrix" category.

Been thinking about the non-headless mode of pantalaimon. Wouldn't this mode fit better in something like home-manager? Or are there examples in nixpkgs of user-services supporting the use-case of multiple users logging in, starting a service and not conflicting with each other (ports, directories, ...)?

Then we could add a note to the pantalaimon service definition pointing to home-manager (and maybe rename it to `services.pantalaimon-headless` to make it obvious) and remove the `user`/`group` options and make the generated services always use `DynamicUser=true`.

jojosch commented:
Also added a test for mjolnir and formatted the tests with `nixpkgs-fmt`. 

The test creates a homeserver, then a mjolnir instance with pantalaimon and finally a client which registers, creates a moderator room and invites mjolnir to that room. The test is successful when mjolnir logs its "Startup complete. Now monitoring rooms" message.

symphorien commented:
> Been thinking about the non-headless mode of pantalaimon. Wouldn't this mode fit better in something like home-manager? Or are there examples in nixpkgs of user-services supporting the use-case of multiple users logging in, starting a service and not conflicting with each other (ports, directories, ...)?

Probably hm is a better option

> Then we could add a note to the pantalaimon service definition pointing to home-manager (and maybe rename it to services.pantalaimon-headless to make it obvious) and remove the user/group options and make the generated services always use DynamicUser=true.

Renaming the modules sounds good, and adding a pointer to hm as well.

sumnerevans commented:
I definitely think feels weird having the GUI option enabled by default in nixpkgs.

sumnerevans commented on nixos/modules/services/matrix/mjolnir.nix:
```diff
@@ -0,0 +1,243 @@
+in
+{
+  options.services.mjolnir = {
+    enable = mkEnableOption "mjolnir";
```
```suggestion
    enable = mkEnableOption "Mjolnir, a moderation tool for Matrix";
```

sumnerevans commented on nixos/modules/services/matrix/mjolnir.nix:
```diff
@@ -0,0 +1,243 @@
+
+    homeserverUrl = mkOption {
+      type = types.str;
+      default = "https://matrix.org";
+      description = ''
+        Where the homeserver is located (client-server URL).
+
+        When using <literal>pantalaimon</literal>, this option will become the homeserver to which <literal>pantalaimon</literal> connects.
+        The listen address of <literal>pantalaimon</literal> will become the <literal>homeserverUrl</literal> of <literal>pantalaimon</literal>.
+      '';
```
I'm not sure what the best way to do this is, but I think it's weird to have this option totally ignored when `pantalaimon` is enabled.

```suggestion
      default = if cfg.pantalaimon.enable then
        "http://${cfg.pantalaimon.options.listenAddress}:${toString cfg.pantalaimon.options.listenPort}"
      else
        "https://matrix.org";
      defaultText = ''
        if <literal>cfg.pantalaimon.enable</literal> is <literal>true</literal>, then the pantalaimon
        listen address and port. Otherwise, <literal>"htps://matrix.org"</literal>
      '';
      description = ''
        Where the homeserver is located (client-server URL).

        When using <literal>pantalaimon</literal>, this option will become the homeserver to which <literal>pantalaimon</literal> connects.
        The listen address of <literal>pantalaimon</literal> will become the <literal>homeserverUrl</literal> of <literal>pantalaimon</literal>.
      '';
```

jojosch replied:
This option isn't ignored when `pantalaimon.enable == true`:

1. If `services.mjolnir.pantalaimon.enable == false`: just use the value as `homeserverUrl` in the mjolnir config. Don't care about pantalaimon.
2. If `services.mjolnir.pantalaimon.enable == true`: "copy" the value to `services.pantalaimon-headless."mjolnir".homeserver` and set `http://<pantalaimonListen>` as the new `homeserverUrl`.

Maybe a bit "magical" - I found a typo in the previous description (pantalaimon -> mjolnir in the last line, last &lt;literal&gt;), maybe now it's clearer.

sumnerevans commented on nixos/modules/services/matrix/pantalaimon-options.nix:
```diff
@@ -0,0 +1,118 @@
+
+    notifications = mkOption {
+      type = types.bool;
+      default = true;
```
If you decide to move the non-headless version to hm, this should be false by default I think.

Or this default should be dependent on the `headless` option

jojosch replied:
Currently in progress of changing it, I've removed the `headless` option and the `notifications`, `useKeyring` and `ignoreVerifications` and set the values in the generated config in `mkConfigFile` for by-default headless operation like 

```nix
{}: recursiveUpdate {
# ...
    SSL = instanceConfig.ssl;

    # Set some settings to prevent user interaction for headless operation
    IgnoreVerification = true;
    Notifications = false;
    UseKeyring = false;
} instanceConfig.extraSettings
```

which can be overridden with the `extraSettings` attribute, if needed.

sumnerevans commented on nixos/modules/services/matrix/pantalaimon-options.nix:
```diff
@@ -0,0 +1,118 @@
+      type = types.path;
+      default = "/var/lib/pantalaimon-${name}";
+      description = ''
+        The directory where <literal>pantalaimon</literal> should store e.g. the database file in.
```
```suggestion
        The directory where <literal>pantalaimon</literal> should store its state such as the database file.
```

jojosch committed:
nixos/pantalaimon: init

jojosch committed:
nixos/pantalaimon: add test

Co-authored-by: sumnerevans <me@sumnerevans.com>

jojosch committed:
nixos/mjolnir: init

jojosch committed:
nixos/mjolnir: add test

jojosch commented:
rebased and fixed the merge conflict after my PR #125333

sumnerevans commented on nixos/modules/services/matrix/pantalaimon-options.nix:
```diff
@@ -0,0 +1,69 @@
+    };
+
+    homeserver = mkOption {
+      type = types.str;
```
```suggestion
      type = types.str;
      example = "https://matrix.org";
```

sumnerevans commented on nixos/modules/services/matrix/pantalaimon-options.nix:
```diff
@@ -0,0 +1,69 @@
+      type = types.bool;
+      default = true;
+      description = ''
+        A boolean that decides if SSL verification should be enabled for
+        outgoing connections to the homeserver.
```
```suggestion
        Whether or not SSL verification should be enabled for outgoing
        connections to the homeserver.
```

sumnerevans reviewed (approved):
Thanks for doing this! It looks good to me.

symphorien commented:
I won't have time to review, sorry

grahamc commented on nixos/modules/services/matrix/mjolnir.nix:
```diff
@@ -0,0 +1,239 @@
+
+      # merge all config files into one, overriding settings of the previous one with the next config
+      # e.g. "eval-all 'select(fileIndex == 0) * select(fileIndex == 1)' filea.yaml fileb.yaml" will merge filea.yaml with fileb.yaml
+      ${pkgs.yq-go}/bin/yq eval-all -P '${yqEvalStr}' ${yqEvalArgs} > ${cfg.dataPath}/config/default.yaml
```
If yq fails this script exits 0, letting the service continue. This should probably use `set -e` to avoid that.

jojosch replied:
added some error handling

grahamc commented on nixos/modules/services/matrix/mjolnir.nix:
```diff
@@ -0,0 +1,239 @@
+    ''
+      # mjolnir will try to load a config from "./config/default.yaml" in the working directory
+      # -> let's place the generated config there
+      mkdir -p ${cfg.dataPath}/config
```
It looks like the mjolnir config is created with some somewhat generous permissions:

```
[root@kif:~]# stat /var/lib/mjolnir/config/default.yaml
  File: /var/lib/mjolnir/config/default.yaml
  Size: 6550      	Blocks: 10         IO Block: 6656   regular file
Device: 1bh/27d	Inode: 49088       Links: 1
Access: (0644/-rw-r--r--)  Uid: (  990/ mjolnir)   Gid: (  989/ mjolnir)
Access: 2021-06-14 15:07:38.557290877 +0000
Modify: 2021-06-14 15:12:39.376004070 +0000
Change: 2021-06-14 15:12:39.376004070 +0000
 Birth: -

```

however an unprivileged user can't access it due to the parent directory:

```
[root@kif:~]# stat /var/lib/mjolnir/
  File: /var/lib/mjolnir/
  Size: 4         	Blocks: 1          IO Block: 131072 directory
Device: 1bh/27d	Inode: 264677      Links: 3
Access: (0700/drwx------)  Uid: (  990/ mjolnir)   Gid: (  989/ mjolnir)
Access: 2021-06-14 15:07:38.538291044 +0000
Modify: 2021-06-14 15:12:02.775251291 +0000
Change: 2021-06-14 15:12:02.775251291 +0000
 Birth: -
```

but I wonder if we should drop the umask anyway.

jojosch replied:
added an more restrictive `umask`

grahamc commented on nixos/modules/services/matrix/mjolnir.nix:
```diff
@@ -0,0 +1,239 @@
+{ config, lib, pkgs, ... }:
```
The service is in a restart loop by default when connecting to an EMS hosted Matrix:

```
Jun 14 15:14:45 kif systemd[1]: Started mjolnir - a moderation tool for Matrix.
Jun 14 15:14:45 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:45 GMT [INFO] [index] Starting bot...
Jun 14 15:14:46 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:46 GMT [INFO] [index] Resolving protected rooms...
Jun 14 15:14:46 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:46 GMT [INFO] [index] Resolving management room...
Jun 14 15:14:47 kif mjolnir[22947]: Mon, 14 Jun 2021 15:14:47 GMT [ERROR] [MatrixHttpClient (REQ-4)] {
Jun 14 15:14:47 kif mjolnir[22947]:   consent_uri: 'https://xxx.ems.host/_matrix/consent?u=xxx&h=xxx',
Jun 14 15:14:47 kif mjolnir[22947]:   errcode: 'M_CONSENT_NOT_GIVEN',
Jun 14 15:14:47 kif mjolnir[22947]:   error: 'To continue using this homeserver you must review and agree to our terms and conditions at https://xxx.ems.host/_matrix/consent?u=xxx&h=xxx.'
Jun 14 15:14:47 kif mjolnir[22947]: }
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) UnhandledPromiseRejectionWarning: #<IncomingMessage>
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Jun 14 15:14:47 kif mjolnir[22947]: (node:22947) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Jun 14 15:14:47 kif systemd[1]: mjolnir.service: Succeeded.
Jun 14 15:14:47 kif systemd[1]: mjolnir.service: Consumed 495ms CPU time, received 3.1K IP traffic, sent 1.9K IP traffic.
```

I'm a bit surprised it is "Succeeded" instead of failed here, but perhaps this is a bug in mjolnir?

grahamc replied:
After consenting, I get another error that kills the bot but exits with success:

```
Jun 14 15:17:15 kif mjolnir[29663]: Mon, 14 Jun 2021 15:17:15 GMT [ERROR] [MatrixHttpClient (REQ-4)] { errcode: 'M_FORBIDDEN', error: 'You are not invited to this room.' }
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) UnhandledPromiseRejectionWarning: #<IncomingMessage>
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Jun 14 15:17:15 kif mjolnir[29663]: (node:29663) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

jojosch replied:
Had something similar happen when starting up mjolnir and pantalaimon wasn't ready to serve requests yet. The mjolnir process exited with status 0 as well (https://github.com/NixOS/nixpkgs/pull/123896#discussion_r637407051).

Looking at the node warning however, it seems like the mjolnir code doesn't catch that promise rejection. Will try to identify where the check is missing and try to create a patch upstream.

jojosch replied:
Added an upstream PR catching these errors and setting a non-zero exit code. Also added that change as a temporary patch to the mjolnir package.

grahamc reviewed (commented):
- [ ] Needs changelog docs, perhaps noting the issue with the consent error and startup looping

ethancedwards8 commented:
Result of `nixpkgs-review pr 123896` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pantalaimon-headless</li>
  </ul>
</details>

jojosch committed:
mjolnir: 0.1.17 -> 0.1.18

sumnerevans commented:
There's a new release of mjolnir: https://github.com/matrix-org/mjolnir/releases/tag/v0.1.18

Also, what needs to be done here to get this pushed forward?

jojosch commented:
> There's a new release of mjolnir: https://github.com/matrix-org/mjolnir/releases/tag/v0.1.18

I've updated locally and am currently working on a litte documentation on how to setup everything.

jojosch committed:
nixos/mjolnir: add documentation

jezcope commented:
Great to see this being developed, thanks! How reasonable would it be to adapt the module to enable running multiple instances of the bot? I help out with moderation of several communities and while they might want to share banlists under some circumstances it would be preferable to have a separate moderation bot for each community.

jojosch committed:
mjolnir: 0.1.18 -> 1.1.20

jojosch commented:
Rebased, fixed the conflict in `pantalaimon/default.nix` and `all-tests.nix` and updated mjolnir to the latest release https://github.com/matrix-org/mjolnir/releases/tag/v1.1.20. This release also contains the patch to set an exit-code when encountering an error.

https://github.com/matrix-org/mjolnir/releases/tag/v0.1.19 was released in the meantime.

I've run the mjolnir tests and they succeed.

-----------------

@jezcope: I will try to add multiple instances of mjolnir (like the pantalaimon instances).

grahamc committed:
Merge branch 'master' into mjolnir-init

grahamc committed:
nixos: mjolnir: literalExample -> literalExpression

grahamc commented:
I meant to merge this months ago when I first started deploying mjolnir for the Matrix space. Thanks for this PR.
